src/hotspot/share/opto/rangeinference.hpp

```diff
@@ -0,0 +1,179 @@
+
+public:
+  U _zeros;
+  U _ones;
+
+  bool is_satisfied_by(U v) const {
+    return (v & _zeros) == 0 && (v & _ones) == _ones;
+  }
+};
+
+// All the information needed to construct a TypeInt/TypeLong, the constraints
+// here may be arbitrary and need to be canonicalized to construct a
+// TypeInt/TypeLong
+template <class S, class U>
+class TypeIntPrototype {
+public:
+  static_assert(std::is_signed<S>::value, "");
+  static_assert(std::is_unsigned<U>::value, "");
+  static_assert(sizeof(S) == sizeof(U), "");
+
+  RangeInt<S> _srange;
+  RangeInt<U> _urange;
+  KnownBits<U> _bits;
+
+private:
+  friend class TypeInt;
+  friend class TypeLong;
+
+  template <class T1, class T2>
+  friend void test_canonicalize_constraints_simple();
+
+  template <class T1, class T2>
+  friend void test_canonicalize_constraints_random();
+
+  // A canonicalized version of a TypeIntPrototype, if the prototype represents
+  // an empty type, _present is false, otherwise, _data is canonical
+  class CanonicalizedTypeIntPrototype {
+  public:
+    bool _present; // whether this is an empty set
+    TypeIntPrototype<S, U> _data;
+
+    static CanonicalizedTypeIntPrototype make_empty() {
+      return {false, {}};
+    }
+  };
+
+  CanonicalizedTypeIntPrototype canonicalize_constraints() const;
+  int normalize_widen(int w) const;
+#ifdef ASSERT
+  bool contains(S v) const;
+  void verify_constraints() const;
+#endif // ASSERT
+};
+
+// Various helper functions for TypeInt/TypeLong operations
+class TypeIntHelper {
+public:
+  // Calculate the cardinality of a TypeInt/TypeLong ignoring the bits
+  // constraints, the result is tuned down by 1 to ensure the bottom type is
+  // correctly calculated
+  template <class S, class U>
+  static U cardinality_from_bounds(const RangeInt<S>& srange, const RangeInt<U>& urange) {
+    static_assert(std::is_signed<S>::value, "");
+    static_assert(std::is_unsigned<U>::value, "");
+    static_assert(sizeof(S) == sizeof(U), "");
+
+    if (U(srange._lo) == urange._lo) {
+      return urange._hi - urange._lo;
+    }
+
+    return (urange._hi - U(srange._lo)) + (U(srange._hi) - urange._lo) + 1;
+  }
+
+  template <class CT, class S, class U>
+  static const Type* int_type_xmeet(const CT* i1, const Type* t2, const Type* (*make)(const TypeIntPrototype<S, U>&, int, bool), bool dual);
+
+  template <class CT>
+  static bool int_type_equal(const CT* t1, const CT* t2) {
+    return t1->_lo == t2->_lo && t1->_hi == t2->_hi && t1->_ulo == t2->_ulo && t1->_uhi == t2->_uhi &&
+          t1->_bits._zeros == t2->_bits._zeros && t1->_bits._ones == t2->_bits._ones;
+  }
+
+  template <class CT>
+  static bool int_type_subset(const CT* super, const CT* sub) {
+    return super->_lo <= sub->_lo && super->_hi >= sub->_hi && super->_ulo <= sub->_ulo && super->_uhi >= sub->_uhi &&
+          (super->_bits._zeros &~ sub->_bits._zeros) == 0 && (super->_bits._ones &~ sub->_bits._ones) == 0;
+  }
```

Comment:
why are these not member methods of `CT`? It also looks like this could be split for the components - at least the `_bits` could be a member method of `KnownBits`.

Having it more modular would make it easier to read and understand.

Reply:
They are helper methods that get called from `CT`, the main reason is that `TypeInt` and `TypeLong` are not templates so implementing these would be a huge duplication. This class is used merely to help reduce this duplication.

Reply:
Ah fair enough. Ok.